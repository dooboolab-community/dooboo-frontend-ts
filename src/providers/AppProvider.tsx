import React, { useReducer } from 'react';
import { AppContext } from '../contexts';
import { IUser, ILocale } from '../types';
import STRINGS from '../../STRINGS';

const AppConsumer = AppContext.Consumer;

interface IAction {
  type: 'reset-user' | 'set-user';
  payload: any;
}

interface IProps {
  children?: any;
}

export interface IState {
  user: IUser;
  locale?: ILocale;
}

const initialState: IState = {
  user: {
    displayName: '',
    age: 0,
    job: '',
  },
};

const reducer = (state: IState, action: IAction) => {
  switch (action.type) {
    case 'reset-user':
      return { ...state, user: initialState.user };
    case 'set-user':
      return { ...state, user: action.payload };
    default:
      return null;
  }
};

const AppProvider = (props: IProps) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const value = { state, dispatch };
  const ContextProvider = AppContext.Provider;

  return (
    <ContextProvider value={value}>
      {props.children}
    </ContextProvider>
  );
};

export { AppConsumer, AppProvider, AppContext };
