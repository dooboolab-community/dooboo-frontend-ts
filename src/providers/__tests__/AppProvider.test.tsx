import * as React from 'react';

import { AppProvider, useAppContext } from '../AppProvider';
import {
  RenderResult,
  act,
  fireEvent,
  render,
} from '@testing-library/react';

const FakeChild = (): React.ReactElement => {
  const { state, resetUser, callDefault } = useAppContext();

  return (
    <div>
      <div data-testid="TEXT">{JSON.stringify(state, null, 2)}</div>
      <button
        data-testid="BUTTON"
        onClick={(): void => {
          resetUser();
        }}
      />
      <button
        data-testid="BUTTON_NOT_VALID"
        onClick={(): void => {
          callDefault();
        }}
      />
    </div>
  );
};

describe('[AppProvider] rendering test', () => {
  let testingLib: RenderResult;
  const component = (
    <AppProvider>
      <FakeChild />
    </AppProvider>
  );

  it('component and snapshot matches', () => {
    testingLib = render(component);
    const { baseElement } = testingLib;
    expect(baseElement).toMatchSnapshot();
    expect(baseElement).toBeTruthy();
  });

  it('should call [resetUser] when button pressed', () => {
    testingLib = render(component);
    const btn = testingLib.queryByTestId('BUTTON');
    act(() => {
      fireEvent.click(btn);
    });
  });

  it('should call [default] when button pressed', () => {
    testingLib = render(component);
    const btn = testingLib.queryByTestId('BUTTON_NOT_VALID');
    act(() => {
      fireEvent.click(btn);
    });
  });
});

// todo: add interaction test
