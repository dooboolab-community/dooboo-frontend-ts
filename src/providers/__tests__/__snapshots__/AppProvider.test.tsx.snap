// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[AppProvider] rendering test component and snapshot matches 1`] = `
<body>
  <div>
    <div>
      <div
        data-testid="TEXT"
      >
        {
  "user": {
    "displayName": "",
    "age": 0,
    "job": ""
  }
}
      </div>
      <button
        data-testid="BUTTON"
      />
      <button
        data-testid="BUTTON_NOT_VALID"
      />
    </div>
  </div>
</body>
`;
